metagpt/roles/role.py

```diff
@@ -533,9 +534,14 @@ async def act(self) -> ActionOutput:
         return ActionOutput(content=msg.content, instruct_content=msg.instruct_content)
 
     @property
-    def first_action(self) -> str:
-        """AgentStore uses this attribute to display to the user what actions the current role should take."""
-        # FIXME: this is a hack, we should not use the first action to represent the todo
+    def action_description(self) -> str:
+        """
+        AgentStore uses this attribute to display to the user what actions the current role should take.
```

Comment:
This naming loses the meaning of the first action, the handling is implicit and needs to be commented in the function

Reply:
The problem here is that action_description doesn't express what strategy it wants to use to select actions, and the code lacks corresponding comments. In particular, it seems that there should be a unified "function that is about to execute an action" before converting it to str, otherwise developers will be confused about which action to execute and where

Reply:
`Role` provides the default property, and this property can be overridden by children classes.
Take `Engineer` as example:
```
class Engineer(Role):
    @property
    def action_description(self) -> str:
        """AgentStore uses this attribute to display to the user what actions the current role should take."""
        return self.next_todo_action
        
async def _act(self) -> Message | None:
        """Determines the mode of action based on whether code review is used."""
        if self.rc.todo is None:
            return None
        if isinstance(self.rc.todo, WriteCode):
            self.next_todo_action = any_to_name(SummarizeCode)
            return await self._act_write_code()
        if isinstance(self.rc.todo, SummarizeCode):
            self.next_todo_action = any_to_name(WriteCode)
            return await self._act_summarize()
        return None
```